crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -57,17 +67,84 @@ where
         self.inner.install_filter(FilterKind::PendingTransaction).await
     }
 
-    async fn filter_changes(&self, _id: SubscriptionId<'_>) -> RpcResult<FilterChanges> {
-        todo!()
+    async fn filter_changes(&self, id: SubscriptionId<'_>) -> RpcResult<FilterChanges> {
+        let id = id.into_owned();
+        let info = self.inner.client.chain_info().to_rpc_result()?;
+        let best_number = info.best_number;
+
+        let (start_block, kind) = {
+            let mut filters = self.inner.active_filters.inner.lock().await;
+            let mut filter = filters.get_mut(&id).ok_or_else(|| FilterError::FilterNotFound(id))?;
+
+            // update filter
+            // we fetch all changes from [filter.block..best_block], so we advance the filter's
+            // block to `best_block +1`
+            let mut block = best_number + 1;
+            std::mem::swap(&mut filter.block, &mut block);
+            filter.last_poll_timestamp = Instant::now();
+
+            (block, filter.kind.clone())
+        };
+
+        match kind {
+            FilterKind::PendingTransaction => {
+                return Err(internal_rpc_err("method not implemented"))
+            }
```

Comment:
Is it even possible to do get filter changes for txs? maybe something we should disallow at the type level

Reply:
I'm actually not sure if anyone supports this as filter

Reply:
ethers-rs supports it via the transaction stream in ethers-providers

Reply:
oh no this convo is about which filters you can poll with eth_filterChanges and pending txs are not a filter they're a subscription

Reply:
ya, I know. I mean that ethers-rs has special-case code written for working with the eth_filterChanges for pending txns. Helps wrap it in a struct that retrieves each TX as it sees the hash in the filter changes

https://github.com/gakonst/ethers-rs/blob/master/ethers-providers/src/stream.rs#L127-L140